src/LinearSolverTypes.f90

```diff
@@ -1420,43 +1494,55 @@ SUBROUTINE setConv_LinearSolverType_Iterative(solver,normType_in,  &
 !>
 !> This subroutine gets the residual after completion of the iterative solver
 !>
-SUBROUTINE getResidual_LinearSolverType_Iterative(solver,resid)
+SUBROUTINE getResidualVec_LinearSolverType_Iterative(solver,resid,nIters)
+  CHARACTER(LEN=*),PARAMETER :: myName='getResidualVec_LinearSolverType_Iterative'
   CLASS(LinearSolverType_Iterative),INTENT(INOUT) :: solver
-  TYPE(RealVectorType),INTENT(INOUT) :: resid
+  CLASS(VectorType),INTENT(INOUT) :: resid
+  INTEGER(SIK),INTENT(OUT),OPTIONAL :: nIters
   !input check
-  IF(solver%isInit .AND. ASSOCIATED(solver%b) .AND. ASSOCIATED(solver%A) &
-      .AND. ASSOCIATED(solver%X) .AND. resid%n > 0) THEN
-    !Written assuming A is not decomposed.  Which is accurate, the correct
-    !solve function will contain the decomposed A.
-    IF(resid%n == solver%b%n) THEN
-#ifdef HAVE_MKL
-      !not yet implemented
-      SELECTTYPE(b => solver%b); TYPE IS(RealVectorType)
-        resid%b=-b%b
-      ENDSELECT
-      CALL BLAS_matvec(THISMATRIX=solver%A,X=solver%X,Y=resid)
-#else
-      !perform calculations using the BLAS system (intrinsic to Futility or
-      !TPL, defined by #HAVE_BLAS)
-      SELECTTYPE(b => solver%b); TYPE IS(RealVectorType)
-        resid%b=-b%b
-      ENDSELECT
-      CALL BLAS_matvec(THISMATRIX=solver%A,X=solver%X,Y=resid)
-#endif
-    ENDIF
+  REQUIRE(resid%isInit)
+  REQUIRE(solver%isInit)
+  REQUIRE(ASSOCIATED(solver%b))
+  REQUIRE(solver%b%isInit)
+  REQUIRE(ASSOCIATED(solver%A))
+  REQUIRE(solver%A%isInit)
+  REQUIRE(ASSOCIATED(solver%x))
+  REQUIRE(solver%x%isInit)
+  REQUIRE(resid%isInit)
+  !Written assuming A is not decomposed.  Which is accurate, the correct
+  !solve function will contain the decomposed A.
+
+  !perform calculations using the BLAS system (intrinsic to Futility or
+  !TPL, defined by #HAVE_BLAS)
+  ! MKL May need something special but is not implemented at the moment.
+  IF (resid%isInit .AND. solver%isInit) THEN
+    CALL BLAS_copy(solver%b,resid)
+    CALL BLAS_matvec(thisMatrix=solver%A,alpha=-1.0_SRK,x=solver%x,beta=1.0_SRK,y=resid)
+    IF (PRESENT(nIters)) nIters = solver%iters
+  ELSE
+    CALL eLinearSolverType%raiseError('Incorrect call to '// &
+       modName//'::'//myName//' - Residual or solver not initialized.')
   ENDIF
-ENDSUBROUTINE getResidual_LinearSolverType_Iterative
-!
-!-------------------------------------------------------------------------------
-!>
-SUBROUTINE getIterResidual_LinearSolverType_Iterative(solver,niters,resid)
+
+ENDSUBROUTINE getResidualVec_LinearSolverType_Iterative
+
+SUBROUTINE getResidualNorm_LinearSolverType_Iterative(solver,resid,nIters)
```

Comment:
No subroutine description

Reply:
Oops, thought I caught this one on the last commit. I'll add the brief.